### DISCUSSION TITLE
Is there any option to put string of F functions into script from outside

### DISCUSSION THREAD

**Discussion by eudjinnl** [nornir-automation/nornir#833] [Q&A]
Let's say I have a script which sends commands to devices.
today I need to filter inventory like:
nornir_object.filter(F(site='site1") & F(model__contains = "3850"))
the next day it can be
nornir_object.filter(F(site='site2") & F(model__contains = "2960"))

So now I open the script, write the right filter string, execute scrypt.

Maybe there is an option to send filter string to the script fro outside?
like a string like
'F(site='site2") & F(model__contains = "2960")'
from separate file or something like this

**Comment by jeremystretch** [comment]
You can just reference variables when calling `F()`, after populating those variables elsewhere. For example:

```python
import sys

site_name = sys.argv[1]
model_name = sys.argv[2]

nornir_object.filter(F(site=site_name) & F(model__contains=model_name))
```

**Reply by eudjinnl** [reply]
In this case filter string is predefined only variables change.
But I'm looking for a solution when filter string changes.
like today i need
nornir_object.filter(F(site='site2") & F(model__contains = "2960"))
next time i need
nornir_object.filter((F(site='site2") | F(site='site1")) & F(model__contains = "2960"))
next time
nornir_object.filter((F(site='site2") | F(site='site1")) & (F(model__contains = "2960") | F(model__contains = "3850")))

and so on

as for now I've made a small and simple  script only with filter function and change it as I need.
big script which works with devices imports filter function from small script and thus I get filtered nornit object.

so I'm looking for an option to edit filter string outside script file and insert it into script.

**Reply by ubaumann** [reply]
There is an open pull request, #603, but I assume no one has time to finalize it.  

I usually have a couple of filter parameters in my scripts and filter the inventory if a parameter is given. I keep it as simple as possible. Sometimes it is easier to filter multiple times and not have complex solutions. 

```python
if tags:
    nr = nr.filter(F(tags__any=tags))
if host_filter:
    nr = nr.filter(F(name__contains=host_filter))
```

**Reply by dbarrosop** [reply]
To be fair, that PR was 99% done, only bits missing were docs and feedback from people. My time right now is kinda awful but if someone wants to takeover and help with the docs it should be pretty close to be merged.

**Reply by eudjinnl** [reply]
I have a look on PR 603 and it feels like this is the thing I'm looking for.

I'm quite a noob in python and git but I'm ready to try to use this function and give a feedback. 
Am I right that I have to make a fork into my github account?

**Reply by eudjinnl** [reply]
I've made a couple of quick tests with parser function. Everything looks fine for time being. This is the thing I asked for.
I've faced  some exceptions but I dont know is it a feature or a bug.
Should I continue futher discussion here of move it to PR 603 discussion?

**Reply by dbarrosop** [reply]
It's probably better to do it in that PR.
